docs: proposal for switching gateways to a central agent (#1855)
kyma-project/telemetry-manager: docs/contributor/arch/019-switch-from-gateways-to-a-central-agent.md

- Every instance caches the whole cluster state

  Every instance of a gateway must be able to enrich the data of every Pod and with that need to cache all Pod metadata.

- Istio mandatory for load balancing

  Pushing OTLP data from an application to the gateway running with replicas requires load balancing. Unfortunately, a Kubernetes service is acting on L4 while GRPC/HTTP2 is on L7, so an additional load balancer is needed. That can be easily solved by using Istio, which becomes a requirement for applications. However, some applications don't use Istio (StatefulSets often don't) but want to provide OTLP data.

- Istio is mandatory for mTLS

  App to Gateway communication is across nodes and with that should use mTLS, so Istio needs to be used which sometimes is not possible. Also, Istio causes additional overhead like noise filtering and collecting observability data about the service mesh should not rely on the service mesh.

- Cumulativetodelta transformation not supported

>   This transformation is mandatory for Dynatrace support and requires that data from a Pod gets processed always by the same instance.

Reviewer: Yes, it's this section in particular https://docs.dynatrace.com/docs/ingest-from/opentelemetry/collector/configuration#delta-metrics

>   Cumulativetodelta transformation is mandatory for Dynatrace support and requires that data from a Pod is always processed by the same instance.

Author: ```suggestion
  The usage of the `cumulativetodelta` processor is mandatory for Dynatrace support and requires that data from a Pod is always processed by the same instance.
```

- General misconception

>   Gateways usually run outside the cluster as a first entry point. The cluster should get rid of the data as fast as possible, because persistence cannot be provided here (things like Kafka would run outside again). Try to compensate for short-lived hiccups. Otherwise, point back the backpressure to the client and do not try to solve the big picture.

Author: ```suggestion
  The wish of having a dispatching solution inside a cluster, with a configurable persistent buffer and well-defined guarantees is not realistic to achieve with simple kubernetes primitives. A queueing system like Kafka will be required and usually will run as a managed solution outside of the cluster. The elements inside the cluster should focus on processing node-local data as fast as possible and delivering the data to the next service which can provide guarantees.
  Here, we should focus on the agent scenario only by enriching data node-local, leveraging the natural scaling by node instances. In backpressure scenarios, it should not be tried to buffer in a persistent way for a longer time, but instead propagate back the backpressure to the application so that it can react to it.
```

What about that?

>   Gateways usually run outside the cluster as a first entry point. The cluster should get rid of the data as fast as possible, because persistence cannot be provided here (things like Kafka would run outside again). Try to compensate for short-lived hiccups. Otherwise, point back the backpressure to the client and do not try to solve the big picture.

Reviewer 2: Icke's suggestion, language-reviewed:

```suggestion
  Using simple Kubernetes primitives, it's unrealistic to create a dispatching solution inside a cluster, with a configurable persistent buffer and well-defined guarantees. A queueing system like Kafka will be required and usually will run as a managed solution outside of the cluster. The elements inside the cluster should focus on processing node-local data as fast as possible and delivering the data to the next service, which can provide guarantees.
  Here, we should focus only on the agent scenario: By enriching data node-locally and using the natural scaling done by node instances. In backpressure scenarios, it should not be tried to buffer in a persistent way for a longer time, but instead, propagate the backpressure back to the application so that it can react to it.
```

## Proposal

> Most of the drawbacks can be solved by running the gateway logic node-local only. Every instance always processes only data of the local node. With that a natural scalability is given and can be extended by vertical scaling capabilities. The "old" gateways will be running in agent mode but still called "OTLP Gateway". Additionally, there will be a new uniform service `telemetry-otlp` forwarding to the node-local entity using the [internalTrafficPolicy: local](https://kubernetes.io/docs/reference/networking/virtual-ips/#internal-traffic-policy) setting. Existing services can stay compatible using the same approach.

Reviewer 2: ```suggestion
Most of the drawbacks can be solved by running the gateway logic node-locally only. Every instance always processes only data of the local node. With that, a natural scalability is given and can be extended by vertical scaling capabilities. The "old" gateways will be running in agent mode but still called "OTLP Gateway". Additionally, there will be a new uniform service `telemetry-otlp` forwarding to the node-local entity using the [internalTrafficPolicy: local](https://kubernetes.io/docs/reference/networking/virtual-ips/#internal-traffic-policy) setting. Existing services can stay compatible using the same approach.
```

> Most of the drawbacks can be solved by running the gateway logic node-local only. Every instance always processes only data of the local node. With that a natural scalability is given and can be extended by vertical scaling capabilities. The "old" gateways will be running in agent mode but still called "OTLP Gateway". Additionally, there will be a new uniform service `telemetry-otlp` forwarding to the node-local entity using the [internalTrafficPolicy: local](https://kubernetes.io/docs/reference/networking/virtual-ips/#internal-traffic-policy) setting. Existing services can stay compatible using the same approach.

Reviewer 2: 'The "old" gateways will be running in agent mode but still called "OTLP Gateway"' 

Isn't this going to be confusing? I trust you have reasons to keep the old name instead of calling them "OTLP Agent", but I'm concerned about the potential for confusion. 
Are we going to document the components as:

> This is the agent, it's called OTLP Gateway.
> These are the gateways, they are called ...node gateway? ...node-local gateway? They're also based on OTLP btw ;)

Author: It will be hard to rename it as we use the term in our API. A pipeline has a agent and a gateway running.. Maybe we can come up with a completely different term, let's see. I think we don't have to decide now on the name

![arch](./../assets/otlp-gateway-new.drawio.svg)

### Remaining Drawbacks of the Agent Approach  

#### Coupling of Signal Types  

Decoupling of signal types will no longer be possible, which may result in scenarios where logs are dropped due to metric overload. However, with proper hardening, this should not pose a significant issue.  

> Let's compare a **shared OTLP agent setup** with an **agent-per-signal setup** in terms of robustness. The OTel Collector offers two key mechanisms to prevent overload:

Reviewer 3: (optional) You might consider moving this sentence elsewhere, deleting it, or rephrasing it, since a proper comparison of the two approaches is not explicitly provided, just the explanation of some mechanisms and a paragraph that explains how introducing the new approach might change these mechanisms' configurations and behaviours.

Author: ```suggestion
To mitigate the coupling problem, an agent per signal type could be used instead of one shared agent. Let's compare a **shared OTLP agent setup** with an **agent-per-signal setup** in terms of robustness. The OTel Collector offers two key mechanisms to prevent overload:
```

1. **Memory Limiter**  
   - Applied globally, no way to restrict it to a specific pipeline or signal type.

>    - Performs periodic checks of memory usage and will begin refusing data and forcing GC to reduce memory consumption when defined limits have been exceeded. It influenced by:

Reviewer 2: ```suggestion
   - Performs periodic checks of memory usage. When defined limits have been exceeded, it begins refusing data and forcing GC to reduce memory consumption. It is influenced by:
```

>      - Heap allocated object held by the Collector.  It can be the OTLP exporter sending queues or various caches used by different components in the chain (e.g. Kubernetes Attributes Processor informers cache).  

Reviewer 2: ```suggestion
     - Heap-allocated object held by the OTel Collector.  It can be the OTLP exporter sending queues, or various caches used by different components in the chain (like the Kubernetes Attributes Processor informers cache).  
```

>      - Heap allocated object held by the Collector.  It can be the OTLP exporter sending queues or various caches used by different components in the chain (e.g. Kubernetes Attributes Processor informers cache).  

Reviewer 2: Not sure what the following terms mean, we didn't introduce them earlier:

- **GC** - is this a collector? Gateway Collector?
- **Heap allocated object** - is this an object that's heap-allocated? Then I believe we need a hyphen because "heap allocated" is used as an adjective.
- **Kubernetes Attributes Processor informers cache** - is this the cache of the/an informer of the Kubernetes Attributes Processor? Then maybe it should be "Kubernetes Attributes Processor informer's cache", but even clearer: "Cache of the Kubernetes Attributes Processor informer". Or is it the informer cache of the Kubernetes Attributes Processor? Then drop the "s": "Kubernetes Attributes Processor informer cache"

>      - Heap allocated objects not being used and can be reclaimed by GC.

Reviewer 2: ```suggestion
     - Heap-allocated objects that aren't being used, which can be reclaimed by GC.
```
Not sure I got that correctly

Reviewer: You got it right

>    - Memory Limiter can not prevent all kinds of OOM issues. It can only prevent issues caused by ingested telemetry data by refusing it. For example, the issue of the Kubernetes Attributes Processor informer cache consuming excessive memory can not be prevented by Memory Limiter.  Basically, it is a poor man's throttling mechanism that was necessary when using the old Batch Processor. [Read more](./017-fault-tolerant-otel-logging-setup.md).

Reviewer 2: ```suggestion
   - Memory Limiter can not prevent all kinds of OOM issues. It can only prevent issues caused by ingested telemetry data by refusing it. For example, Memory Limiter cannot prevent the issue of the Kubernetes Attributes Processor informer cache consuming excessive memory. Basically, it is a poor man's throttling mechanism that was necessary when using the old Batch Processor. See [ADR-017](./017-fault-tolerant-otel-logging-setup.md).
```

2. **Refusing Data if OTLP Exporter Queue is Full**  
   - The Batch Processor must not be used (the built-in OTLP Exporter batcher does not have this limitation and can still be used).
   - If the queue is full, new data will be refused.  
   - Each pipeline type has its own queue, allowing different queue lengths to support varying quality-of-service (QoS) levels for different signal types.  

> Introducing a shared collector handling OTLP data for all three telemetry types means that only one memory limiter is configured, instead of three separate ones when they are isolated. However, if OTLP exporter queue refusal is correctly set up, the system should never reach the point where the memory limiter starts rejecting data. across all clusters is not solvable by either the old or the new proposed setup.

Reviewer 2: ```suggestion
Introducing a shared collector that handles OTLP data for all three telemetry types means that only one memory limiter is configured, instead of three separate ones when they are isolated. However, if OTLP exporter queue refusal is correctly set up, the system should never reach the point where the memory limiter starts rejecting data. [MISSING WORDS] across all clusters is not solvable by either the old or the new proposed setup.
```

In front of "across", some words seem to be missing. What is it that isn't solvable?
Maybe something like "Data rejection across all..."

Reviewer: Sorry, it should have been " the Kubernetes Attribute Processor memory usage issue" or something (explained above, the one with the informers cache)

#### Downtime During Updates  

> Unlike deployments, the agent cannot be rolled out in a rolling manner, meaning updates will always result in some downtime. However, this should be mitigated by the retry mechanism in the OpenTelemetry SDK (according to our [tests](../../contributor/pocs/otelcol-downtime/otelcol-downtime.md) it's not yet the case).

Reviewer 2: ```suggestion
Unlike deployments, the agent cannot be rolled out in a rolling manner, meaning updates will always result in some downtime. However, this should be mitigated by the retry mechanism in the OpenTelemetry SDK (according to our [tests](../../contributor/pocs/otelcol-downtime/otelcol-downtime.md), it's not yet the case).
```

### Additional Benefits  

#### Persistent Queue Everywhere  

A small persistent buffer can be introduced in all components using the node file system, enhancing resilience and reliability.  

#### No Istio Dependency  

Neither the application nor its components will depend on Istio anymore. However, the metric agent should still support scraping Istio-instrumented endpoints.  

## Conclusion

> The former motivation of the gateway concept turned out to be no longer relevant. Switching to the agent approach solves many problems while introducing very soft drawbacks. The transformation should start immediately :)

Reviewer 2: Exciting times!
